File: backend/src/main/java/com/example/bbmovie/service/payment/vnpay/VNPayAdapter.java
package com.example.bbmovie.service.payment.vnpay;

import com.example.bbmovie.exception.VNPayException;
import com.example.bbmovie.service.payment.PaymentProviderAdapter;
import com.example.bbmovie.service.payment.PaymentStatus;
import com.example.bbmovie.service.payment.dto.PaymentRequest;
import com.example.bbmovie.service.payment.dto.PaymentResponse;
import com.example.bbmovie.service.payment.dto.PaymentVerification;
import com.example.bbmovie.service.payment.dto.RefundResponse;
import com.example.bbmovie.utils.IpAddressUtils;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import lombok.extern.log4j.Log4j2;
import org.apache.http.NameValuePair;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.util.EntityUtils;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.math.BigDecimal;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.stream.Collectors;

import static com.example.bbmovie.service.payment.vnpay.VnPayQueryParams.*;

@Log4j2
@Service("vnpayProvider")
public class VNPayAdapter implements PaymentProviderAdapter {

    @Value("${payment.vnpay.TmnCode}")
    private String tmnCode;

    @Value("${payment.vnpay.HashSecret}")
    private String hashSecret;

    @Override
    public PaymentResponse processPayment(PaymentRequest request, HttpServletRequest httpServletRequest) {
        String orderId = request.getOrderId();
        String amount = request.getAmount().multiply(new BigDecimal(100)).toString();
        String vnpTxnRef = System.currentTimeMillis()
                + UUID.randomUUID().toString().replace("-", "").substring(0, 32);
        String vnpIpAddr = IpAddressUtils.getClientIp(httpServletRequest);
        String vnpCreateDate = new SimpleDateFormat(VnPayConstraint.DATE_FORMAT).format(new Date());

        Map<String, String> vnpParams = new HashMap<>();
        vnpParams.put(VNPAY_VERSION_PARAM, VnPayConstraint.VERSION);
        vnpParams.put(VNPAY_COMMAND_PARAM, VnPayCommand.PAY.getCommand());
        vnpParams.put(VNPAY_TMN_CODE_PARAM, tmnCode);
        vnpParams.put(VNPAY_AMOUNT_PARAM, amount);
        vnpParams.put(VNPAY_CURRENCY_PARAM, VnPayConstraint.ONLY_SUPPORTED_CURRENCY);
        vnpParams.put(VNPAY_TXN_REF_PARAM, vnpTxnRef);
        vnpParams.put(VNPAY_ORDER_INFO_PARAM, "Order " + orderId);
        vnpParams.put(VNPAY_ORDER_TYPE_PARAM, VnPayOrderType.BILL_PAYMENT.getType());
        vnpParams.put(VNPAY_LOCALE_PARAM, VnPayConstraint.ONLY_SUPPORTED_CURRENCY);
        vnpParams.put(VNPAY_RETURN_URL_PARAM, VnPayConstraint.RETURN_URL);
        vnpParams.put(VNPAY_IP_ADDRESS_PARAM, vnpIpAddr);
        vnpParams.put(VNPAY_CREATE_DATE_PARAM, vnpCreateDate);

        String secureHash = generateSecureHash(vnpParams);
        vnpParams.put(VNPAY_SECURE_HASH, secureHash);

        String redirectUrl = VnPayConstraint.PAYMENT_URL + "?" + toQueryString(vnpParams);
        return new PaymentResponse(vnpTxnRef, PaymentStatus.PENDING, redirectUrl);
    }

    @Override
    public PaymentVerification verifyPayment(Map<String, String> paymentData, HttpServletRequest httpServletRequest) {
        String vnpSecureHash = paymentData.get(VNPAY_SECURE_HASH);
        String vnpTxnRef = paymentData.get(VNPAY_TXN_REF_PARAM);
        paymentData.remove(VNPAY_SECURE_HASH);

        String calculatedHash = generateSecureHash(paymentData);
        boolean isValid = vnpSecureHash.equals(calculatedHash)
                && VnPayTransactionStatus.SUCCESS.getCode().equals(paymentData.get(VNPAY_RESPONSE_CODE));
        return new PaymentVerification(isValid, vnpTxnRef);
    }

    @Override
    public RefundResponse refundPayment(String paymentId, HttpServletRequest httpServletRequest) {
        // Retrieve original transaction details from the database or VNPay API
        // For simplicity, assume transaction details are stored or provided
        // Original transaction ID
        String vnpTransactionNo = getTransactionNo(paymentId); // Fetch from stored data or API
        String vnpTransactionDate = getTransactionDate(paymentId); // Fetch from stored data or API
        String vnpCreateDate = new SimpleDateFormat(VnPayConstraint.DATE_FORMAT).format(new Date());
        String vnpIpAddr = IpAddressUtils.getServerIp();
        String vnpAmount = "10000"; // Example: refund 100 VND (replace with actual amount * 100)

        Map<String, String> vnpParams = new HashMap<>();
        vnpParams.put(VNPAY_VERSION_PARAM, VnPayConstraint.VERSION);
        vnpParams.put(VNPAY_COMMAND_PARAM, VnPayCommand.REFUND.getCommand());
        vnpParams.put(VNPAY_TRANSACTION_TYPE, "02"); // Full refund
        vnpParams.put(VNPAY_TMN_CODE_PARAM, tmnCode);
        vnpParams.put(VNPAY_TXN_REF_PARAM, paymentId);
        vnpParams.put(VNPAY_AMOUNT_PARAM, vnpAmount);
        vnpParams.put(VNPAY_TRANSACTION_NO, vnpTransactionNo);
        vnpParams.put(VNPAY_TRANSACTION_DATE, vnpTransactionDate);
        vnpParams.put(VNPAY_CREATE_DATE_PARAM, vnpCreateDate);
        vnpParams.put(VNPAY_IP_ADDRESS_PARAM, vnpIpAddr);
        vnpParams.put(VNPAY_ORDER_INFO_PARAM, "Refund for transaction " + paymentId);

        String secureHash = generateSecureHash(vnpParams);
        vnpParams.put(VNPAY_SECURE_HASH, secureHash);

        String response = sendRefundRequest(vnpParams);
        assert response != null;
        Map<String, String> responseParams = parseResponse(response);

        String responseCode = responseParams.get(VNPAY_RESPONSE_CODE);
        String refundId = responseParams.get(VNPAY_TRANSACTION_NO);
        return new RefundResponse(
                refundId,
                VnPayTransactionStatus.SUCCESS.getCode().equals(responseCode)
                    ? PaymentStatus.SUCCEEDED.getStatus()
                    : PaymentStatus.FAILED.getStatus()
        );
    }

    public Object queryPayment(String paymentId, HttpServletRequest httpServletRequest) {
        String vnpRequestId = UUID.randomUUID().toString().replace("-", "").substring(0, 32);
        String vnpCreateDate = new SimpleDateFormat(VnPayConstraint.DATE_FORMAT).format(new Date());
        String vnpTransactionDate = getTransactionDate(paymentId); // Fetch this from your DB
        String vnpIpAddr = IpAddressUtils.getClientIp(httpServletRequest);
        String vnpCreateBy = "system"; // Or authenticated admin username

        Map<String, String> params = new HashMap<>();
        params.put("vnp_RequestId", vnpRequestId);
        params.put("vnp_Version", VnPayConstraint.VERSION);
        params.put("vnp_Command", "querydr");
        params.put("vnp_TmnCode", tmnCode);
        params.put("vnp_TxnRef", paymentId);
        params.put("vnp_OrderInfo", "Query for transaction " + paymentId);
        params.put("vnp_TransactionDate", vnpTransactionDate);
        params.put("vnp_CreateBy", vnpCreateBy);
        params.put("vnp_CreateDate", vnpCreateDate);
        params.put("vnp_IpAddr", vnpIpAddr);

        String rawData = String.join("|",
                vnpRequestId,
                VnPayConstraint.VERSION,
                "querydr",
                tmnCode,
                paymentId,
                vnpTransactionDate,
                vnpCreateBy,
                vnpCreateDate,
                vnpIpAddr,
                "Query for transaction " + paymentId
        );

        String secureHash = hmacSHA512(hashSecret, rawData);
        params.put("vnp_SecureHash", secureHash);

        return sendPostRequest(params);
    }


    private String generateSecureHash(Map<String, String> params) {
        List<String> fieldNames = new ArrayList<>(params.keySet());
        Collections.sort(fieldNames);
        StringBuilder sb = new StringBuilder();
        for (String fieldName : fieldNames) {
            sb.append(fieldName).append("=").append(params.get(fieldName)).append("&");
        }
        sb.deleteCharAt(sb.length() - 1);
        return hmacSHA512(hashSecret, sb.toString());
    }

    private String hmacSHA512(final String key, final String data) {
        try {
            if (key == null || data == null) {
                throw new NullPointerException();
            }
            final Mac hmac512 = Mac.getInstance("HmacSHA512");
            byte[] hmacKeyBytes = key.getBytes();
            final SecretKeySpec secretKey = new SecretKeySpec(hmacKeyBytes, "HmacSHA512");
            hmac512.init(secretKey);
            byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
            byte[] result = hmac512.doFinal(dataBytes);
            StringBuilder sb = new StringBuilder(2 * result.length);
            for (byte b : result) {
                sb.append(String.format("%02x", b & 0xff));
            }
            return sb.toString();

        } catch (Exception ex) {
            return "";
        }
    }

    private String toQueryString(Map<String, String> params) {
        return params.entrySet().stream()
                .map(entry -> entry
                        .getKey() + "=" + URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8)
                )
                .collect(Collectors.joining("&"));
    }

    private String sendRefundRequest(Map<String, String> params) throws VNPayException {
        String refundUrl = VnPayConstraint.TRANSACTION_URL;
        try (CloseableHttpClient client = HttpClients.createDefault()) {
            HttpPost post = new HttpPost(refundUrl);
            List<NameValuePair> nvps = new ArrayList<>();
            params.forEach((key, value) -> nvps.add(new BasicNameValuePair(key, value)));
            post.setEntity(new UrlEncodedFormEntity(nvps, StandardCharsets.UTF_8));
            try (CloseableHttpResponse response = client.execute(post)) {
                return EntityUtils.toString(response.getEntity(), StandardCharsets.UTF_8);
            }
        } catch (Exception ex) {
            log.error("Failed to send refund request to VNPay: {}", ex.getMessage(), ex);
            throw new VNPayException("Refund request failed: " + ex.getMessage());
        }
    }

    private String sendPostRequest(Map<String, String> jsonParams) {
        try (CloseableHttpClient client = HttpClients.createDefault()) {
            HttpPost httpPost = new HttpPost(VnPayConstraint.TRANSACTION_URL);
            httpPost.setHeader("Content-Type", "application/json");

            String json = new ObjectMapper().writeValueAsString(jsonParams);
            httpPost.setEntity(new StringEntity(json, StandardCharsets.UTF_8));

            try (CloseableHttpResponse response = client.execute(httpPost)) {
                return EntityUtils.toString(response.getEntity(), StandardCharsets.UTF_8);
            }
        } catch (Exception ex) {
            log.error("Failed to send post request to VNPay: {}", ex.getMessage(), ex);
            throw new VNPayException("Refund request failed: " + ex.getMessage());
        }
    }


    private Map<String, String> parseResponse(String response) {
        Map<String, String> result = new HashMap<>();
        for (String param : response.split("&")) {
            String[] keyValue = param.split("=");
            if (keyValue.length == 2) {
                result.put(keyValue[0], URLDecoder.decode(keyValue[1], StandardCharsets.UTF_8));
            }
        }
        return result;
    }

    // Placeholder: Fetch vnp_TransactionNo from stored data or VNPay API
    private String getTransactionNo(String txnRef) {
        // In practice, query your database or VNPay's transaction query API
        return "VNP" + txnRef; // Example
    }

    // Placeholder: Fetch vnp_TransactionDate from stored data or VNPay API
    private String getTransactionDate(String txnRef) {
        log.info("getTransactionDat {}", txnRef);
        return new SimpleDateFormat(VnPayConstraint.DATE_FORMAT).format(new Date()); // Example
    }
}

File: backend/payment/src/main/kotlin/com/bbmovie/payment/service/payment/vnpay/VNPayAdapter.kt
package com.bbmovie.payment.service.payment.vnpay

import com.bbmovie.payment.entity.PaymentProvider
import com.bbmovie.payment.entity.PaymentStatus
import com.bbmovie.payment.exception.VNPayException
import com.bbmovie.payment.service.payment.PaymentProviderAdapter
import com.bbmovie.payment.service.payment.dto.PaymentRequest
import com.bbmovie.payment.service.payment.dto.PaymentResponse
import com.bbmovie.payment.service.payment.dto.PaymentVerification
import com.bbmovie.payment.service.payment.dto.RefundResponse
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_AMOUNT_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_COMMAND_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_CREATE_DATE_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_CURRENCY_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_IP_ADDRESS_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_LOCALE_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_ORDER_INFO_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_ORDER_TYPE_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_RESPONSE_CODE
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_RETURN_URL_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_SECURE_HASH
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_TMN_CODE_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_TRANSACTION_DATE
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_TRANSACTION_NO
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_TRANSACTION_TYPE
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_TXN_REF_PARAM
import com.bbmovie.payment.service.payment.vnpay.VnPayQueryParams.VNPAY_VERSION_PARAM
import com.bbmovie.payment.utils.IpAddressUtils
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import org.apache.http.client.entity.UrlEncodedFormEntity
import org.apache.http.client.methods.CloseableHttpResponse
import org.apache.http.client.methods.HttpPost
import org.apache.http.entity.StringEntity
import org.apache.http.impl.client.HttpClients
import org.apache.http.message.BasicNameValuePair
import org.apache.http.util.EntityUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.math.BigDecimal
import java.net.URLDecoder
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.text.SimpleDateFormat
import java.util.*
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

@Service("vnpayProvider")
class VNPayAdapter() : PaymentProviderAdapter {

    @Value("\${payment.vnpay.TmnCode}") private lateinit var tmnCode: String
    @Value("\${payment.vnpay.HashSecret}") private lateinit var hashSecret: String

    private val log = LoggerFactory.getLogger(VNPayAdapter::class.java)

    override fun processPayment(
        request: PaymentRequest,
        httpServletRequest: HttpServletRequest
    ): PaymentResponse {
        val orderId = request.orderId
        val amount = request.amount?.multiply(BigDecimal(100)).toString()
        val vnpTxnRef = System.currentTimeMillis().toString() + UUID.randomUUID().toString().replace("-", "").take(32)
        val vnpIpAddr = IpAddressUtils.getClientIp(httpServletRequest)
        val vnpCreateDate = SimpleDateFormat(VnPayConstraint.DATE_FORMAT).format(Date())

        val vnpParams = sortedMapOf(
            VNPAY_VERSION_PARAM to VnPayConstraint.VERSION,
            VNPAY_COMMAND_PARAM to VnPayCommand.PAY.command,
            VNPAY_TMN_CODE_PARAM to tmnCode,
            VNPAY_AMOUNT_PARAM to amount,
            VNPAY_CURRENCY_PARAM to VnPayConstraint.ONLY_SUPPORTED_CURRENCY,
            VNPAY_TXN_REF_PARAM to vnpTxnRef,
            VNPAY_ORDER_INFO_PARAM to "Order $orderId",
            VNPAY_ORDER_TYPE_PARAM to VnPayOrderType.BILL_PAYMENT.type,
            VNPAY_LOCALE_PARAM to VnPayConstraint.ONLY_SUPPORTED_CURRENCY,
            VNPAY_RETURN_URL_PARAM to VnPayConstraint.RETURN_URL,
            VNPAY_IP_ADDRESS_PARAM to vnpIpAddr,
            VNPAY_CREATE_DATE_PARAM to vnpCreateDate
        )

        vnpParams[VNPAY_SECURE_HASH] = generateSecureHash(vnpParams)

        val redirectUrl = VnPayConstraint.PAYMENT_URL + "?" + toQueryString(vnpParams)
        return PaymentResponse(vnpTxnRef, PaymentStatus.PENDING, redirectUrl)
    }

    override fun verifyPayment(
        paymentData: MutableMap<String, String>,
        httpServletRequest: HttpServletRequest
    ): PaymentVerification {
        val data = paymentData.toMutableMap()
        val vnpSecureHash = data.remove(VNPAY_SECURE_HASH)
        val vnpTxnRef = data[VNPAY_TXN_REF_PARAM]

        val calculatedHash = generateSecureHash(data)
        val isValid = vnpSecureHash == calculatedHash &&
                VnPayTransactionStatus.SUCCESS.code == data[VNPAY_RESPONSE_CODE]
        return PaymentVerification(isValid, vnpTxnRef)
    }

    override fun refundPayment(
        paymentId: String,
        httpServletRequest: HttpServletRequest
    ): RefundResponse {
        val vnpTransactionNo = getTransactionNo(paymentId)
        val vnpTransactionDate = getTransactionDate(paymentId)
        val vnpCreateDate = SimpleDateFormat(VnPayConstraint.DATE_FORMAT).format(Date())
        val vnpIpAddr = IpAddressUtils.getPublicServerIp()
        val vnpAmount = "10000" // Example

        val vnpParams = sortedMapOf(
            VNPAY_VERSION_PARAM to VnPayConstraint.VERSION,
            VNPAY_COMMAND_PARAM to VnPayCommand.REFUND.command,
            VNPAY_TRANSACTION_TYPE to "02",
            VNPAY_TMN_CODE_PARAM to tmnCode,
            VNPAY_TXN_REF_PARAM to paymentId,
            VNPAY_AMOUNT_PARAM to vnpAmount,
            VNPAY_TRANSACTION_NO to vnpTransactionNo,
            VNPAY_TRANSACTION_DATE to vnpTransactionDate,
            VNPAY_CREATE_DATE_PARAM to vnpCreateDate,
            VNPAY_IP_ADDRESS_PARAM to vnpIpAddr,
            VNPAY_ORDER_INFO_PARAM to "Refund for transaction $paymentId"
        )

        vnpParams[VNPAY_SECURE_HASH] = generateSecureHash(vnpParams)

        val response = sendRefundRequest(vnpParams) ?: throw VNPayException("Refund request failed")
        val responseParams = parseResponse(response)

        val responseCode = responseParams[VNPAY_RESPONSE_CODE]
        val refundId = responseParams[VNPAY_TRANSACTION_NO]

        val status = if (VnPayTransactionStatus.SUCCESS.code == responseCode)
            PaymentStatus.SUCCEEDED.status
        else PaymentStatus.FAILED.status

        return RefundResponse(refundId ?: "", status)
    }

    override fun getPaymentProviderName(): PaymentProvider {
        return PaymentProvider.VNPAY
    }

    fun queryPayment(paymentId: String, request: HttpServletRequest): Any {
        val vnpRequestId = UUID.randomUUID().toString().replace("-", "").take(32)
        val vnpCreateDate = SimpleDateFormat(VnPayConstraint.DATE_FORMAT).format(Date())
        val vnpTransactionDate = getTransactionDate(paymentId)
        val vnpIpAddr = IpAddressUtils.getClientIp(request)
        val vnpCreateBy = "system"

        val params = mutableMapOf(
            "vnp_RequestId" to vnpRequestId,
            "vnp_Version" to VnPayConstraint.VERSION,
            "vnp_Command" to VnPayCommand.QUERY_DR.command,
            "vnp_TmnCode" to tmnCode,
            "vnp_TxnRef" to paymentId,
            "vnp_OrderInfo" to "Query for transaction $paymentId",
            "vnp_TransactionDate" to vnpTransactionDate,
            "vnp_CreateBy" to vnpCreateBy,
            "vnp_CreateDate" to vnpCreateDate,
            "vnp_IpAddr" to vnpIpAddr
        )

        val rawData = listOf(
            vnpRequestId,
            VnPayConstraint.VERSION,
            VnPayCommand.QUERY_DR.command,
            tmnCode,
            paymentId,
            vnpTransactionDate,
            vnpCreateBy,
            vnpCreateDate,
            vnpIpAddr,
            "Query for transaction $paymentId"
        ).joinToString("|")

        params["vnp_SecureHash"] = hmacSHA512(hashSecret, rawData)

        return sendPostRequest(params) as Any
    }

    private fun generateSecureHash(params: Map<String, String>): String {
        val sortedKeys = params.keys.sorted()
        val sb = StringBuilder()
        for (key in sortedKeys) {
            sb.append("$key=${params[key]}&")
        }
        sb.setLength(sb.length - 1)
        return hmacSHA512(hashSecret, sb.toString())
    }

    private fun hmacSHA512(key: String?, data: String?): String {
        return try {
            requireNotNull(key)
            requireNotNull(data)
            val hmac = Mac.getInstance("HmacSHA512")
            hmac.init(SecretKeySpec(key.toByteArray(), "HmacSHA512"))
            val bytes = hmac.doFinal(data.toByteArray(StandardCharsets.UTF_8))
            bytes.joinToString("") { "%02x".format(it) }
        } catch (ex: Exception) {
            log.error("Failed to generate secure hash: ${ex.message}")
            return ""
        }
    }

    private fun toQueryString(params: Map<String, String>): String {
        return params.entries.joinToString("&") {
            "${it.key}=${URLEncoder.encode(it.value, StandardCharsets.UTF_8)}"
        }
    }

    private fun sendRefundRequest(params: Map<String, String>): String? {
        val client = HttpClients.createDefault()
        val post = HttpPost(VnPayConstraint.TRANSACTION_URL)
        val nvps = params.map { BasicNameValuePair(it.key, it.value) }
        post.entity = UrlEncodedFormEntity(nvps, StandardCharsets.UTF_8)
        return client.use {
            val response: CloseableHttpResponse = it.execute(post)
            EntityUtils.toString(response.entity, StandardCharsets.UTF_8)
        }
    }

    private fun sendPostRequest(jsonParams: Map<String, String>): String? {
        val client = HttpClients.createDefault()
        val post = HttpPost(VnPayConstraint.TRANSACTION_URL)
        post.setHeader("Content-Type", "application/json")
        post.entity = StringEntity(ObjectMapper().writeValueAsString(jsonParams), StandardCharsets.UTF_8)
        return client.use {
            val response = it.execute(post)
            EntityUtils.toString(response.entity, StandardCharsets.UTF_8)
        }
    }

    private fun parseResponse(response: String): Map<String, String> {
        return response.split("&").mapNotNull {
            val (key, value) = it.split("=").let { parts -> parts.getOrNull(0) to parts.getOrNull(1) }
            if (key != null && value != null) key to URLDecoder.decode(value, StandardCharsets.UTF_8) else null
        }.toMap()
    }

    private fun getTransactionNo(txnRef: String): String = "VNP$txnRef"

    private fun getTransactionDate(txnRef: String): String {
        log.info("getTransactionDate {}", txnRef)
        return SimpleDateFormat(VnPayConstraint.DATE_FORMAT).format(Date())
    }
}
